### ISSUE TITLE
The URL helper cannot be used outside request handlers

### ISSUE THREAD

**Issue by lucadealfaro** [web2py/py4web#117]
# Problem: URL() cannot be used outside of request handlers

This same discussion has been put in a PR, but I am repeating here as it involves the URL helper, which is of very general use, so people may want to see/discuss the issue independently of the PR. 

The URL helper can be used _only_ in a controller, not in a declaration. With this I mean that

```
@action("index")
def index():
    u = URL("something")
```

works, but

```
u = URL("something")
@action("index")
def index():
    ...
```

does not.  The problem is that URL relies on the request object to know the application name.

This leads to the problem that one cannot generate a URL outside of a controller.  This is needed for instance in mtable, where one needs to specify the URL for AJAX to the constructor class of mtable.  I am also developing my own version of "grid", which I will be happy to share, and the same problem is present there; I need to do something like: 
```
mygrid = MyGrid(..., callback_url=URL("mygrid", signer=signer))
```

This needs solving, because in py4web, we need to create these classes in the controllers.py code but not in the request handlers, as we cannot register a callback once for each request. 
There are the following solutions to this. 

## Solution 1: make URL application-specific

The most elegant solution would be to define the URL helper as follows: 
```
APPURL = URLFactory("appname")
```

`URLFactory` would take the name of the app as argument, and it would return an object whose `__call__` method acts as the current `URL()` helper.  One would put the above code in `common.py` presumably.  Does this break things elsewhere? 

## Solution 2: enable the URL helper to figure out the app somehow

I have no idea how to do this, but perhaps it can be done. 

## Solution 3: do not sign the application name

Currently, in a request `/app/controller` we sign the entire string `/app/controller`. We could consider signing only `/controller`, so that it would be possible to sign urls also outside of the context of a request.  

From the point of view of security, there is one attack, in which there are `/app1/controller` and `/app2/controller`, and the user is logged _with the same session_ in both app1 and app2.  I do not believe that this is a realistic attack however.  It can already be done with `https://my.host1.com/app/controller` and `https://my.host2.com/app/controller` if one shares sessions.  So I don't think that there is an issue. 

Still, this would lead us to two ways of generating URLs: 

- via the URL helper, when inside a request handler,
- by hand, in a "poor person way", when outside of a request handler. 

This is not super pretty.

**Comment by lucadealfaro** [comment]
Ok, the solution `APPURL = URLFactory("appname")` would break the fact that if you rename an app, 
```
mv apps/app1 apps/app2
```
it would break.  Is there a good solution?

**Comment by lucadealfaro** [comment]
Ok, mental short circuit.  This is not a big deal.  One must simply not use the URL helper when registering callback URLs, but this is not needed, because the signature is not part of the URL.

**Comment by valq7711** [comment]
py4web controllers are just modules, that are imported at server starting, so all that outside the actions executed only once,
but you can
```
def mygrid(URL): return MyGrid(..., callback_url=URL("mygrid", signer=signer))

**Comment by lucadealfaro** [comment]
Yes, above is more or less right.  I closed the bug because it was simply my being confused. 

The point is that sometimes you need to do: 

```
class Grid:
    """This is a prototype class for building paginable grids (tables)
    with content provided server-side."""

    GRID = '<grid load_url="{load_url}"></grid>'

    def __init__(self, session, url=None, signer=None, db=None, auth=None):
        self.url = url or str(uuid.uuid4())
        self.signer = signer or URLSigner(session)
        # Creates an action (an entry point for URL calls),
        # mapped to the api method, that can be used to request pages
        # for the table.
        f = action(self.url, method=["GET"])(self.api)
        args = filter(None, [session, db, auth])
        if args:
            f = action.uses(*args)(f)

    def grid(self):
        """This method returns the element that can be included in the page."""
        return XML(Grid.GRID.format(load_url=URL(self.url, signer=self.signer)))

    def api(self):
        """The API must return the data to fill the table.
        The data is a dictionary, containing:
            - page: <integer>
            - has_more: <boolean>
            - rows: <list of rows (see below)>
        A row is a dictionary, containing:
            - is_header: <boolean>
            - cells: <cells>
        <cells> is a list of dictionaries, containing:
            - text: <text>
            - url: <text> or None
            - is_button: <boolean>
            - el_class: <text> or None (class of element, if needed)
        All the fields except text are optional.
        This is a sample implementation only, to test code.  You should
        over-ride the api method to provide your own input for the table.
        """
        page = request.query.get('page') or 1
        row0 = dict(
            is_header=True,
            cells=[dict(text="Animal"), dict(text="N. paws"), dict(text="Class")])
        row1 = dict(cells=[
            dict(text="Cat"), dict(text="4"),
            dict(text="Mammal", url=URL('mammals/cat'), is_button=True)])
        row2 = dict(cells=[
            dict(text="Dog"), dict(text="4"),
            dict(text="Mammal", url=URL('mammals/dog'), is_button=True)])
        row3 = dict(cells=[
            dict(text="Owl"), dict(text="2"),
            dict(text="Bird", url=URL('bird/owl'), is_button=True)])
        return dict(
            page=int(page),
            has_more=True,
            rows = [row0, row1, row2, row3]
        )
```

and then use this via 

```
# This is the grid server.
samplegrid = grid.Grid(session)

@action('samplegrid', method=['GET'])
@action.uses('grid.html')
def sample_grid():
    """This page generates a sample grid."""
    # We need to instantiate our grid component.
    return dict(grid=samplegrid.grid())
```

So the above registers the callback in `samplegrid.grid()`, which is called in a request handler, and so it can use the URL helper to sign the URL.

**Comment by mdipierro** [comment]
In principle URL could return a lazy object that only runs when we try to
convert to a string (kind of what T does). Is the added complexity worth
the added value?
